lib/core/macros.nim

```diff
@@ -798,6 +798,42 @@ proc newLit*(arg: tuple): NimNode {.compileTime.} =
   for a,b in arg.fieldPairs:
     result.add nnkExprColonExpr.newTree(newIdentNode(a), newLit(b))
 
+proc newLit*(n: NimNode): NimNode {.compileTime, benign.} =
+  ## Convert the AST ``n`` to the code required to generate that AST. Does currently not preserve line information.
+  const LitKinds = nnkLiterals-{nnkNilLit}
+  case n.kind
+  of nnkNilLit:
+    result = newCall(newIdentNode("newNimNode"), newIdentNode("nnkNilLit"))
+  of nnkEmpty:
+    result = newCall(newIdentNode("newEmptyNode"))
+  of nnkIdent:
+    result = nnkCallStrLit.newTree(newIdentNode("ident"), newLit(n.strVal))
+  of nnkSym:
+    assert false, "cannot preserve symbol binding through newLit"
+  of nnkNone:
+    result = newCall(newIdentNode("newNimNode"), newIdentNode("nnkNone"))
+  of nnkCommentStmt:
+    result = newCall("newCommentStmtNode", newLit(n.strVal))
+  of LitKinds:
+    result = newCall(bindSym"newLit", n)
+  else:
+    # some nodes kinds have constructor procs
+    case n.kind
+    of nnkStmtList:
+      result = newCall(newIdentNode"newStmtList")
+    of nnkCall:
+      result = newCall(newIdentNode"newCall")
+    of nnkAsgn:
+      result = newCall(newIdentNode"newAssignment")
+    of nnkDotExpr:
+      result = newCall(newIdentNode"newDotExpr")
+    of nnkExprColonExpr:
+      result = newCall(newIdentNode"newColonExpr")
+    else:
+      result = newCall(nnkDotExpr.newTree(newIdentNode($n.kind), newIdentNode("newTree")))
+    for i in 0 ..< n.len:
+      result.add newLit(n[i])
```

Comment:
```suggestion
    result = newCall(ident"newNimNode", ident"nnkNilLit")
  of nnkEmpty:
    result = newCall(ident"newEmptyNode")
  of nnkIdent:
    result = nnkCallStrLit.newTree(ident"ident", newLit(n.strVal))
  of nnkSym:
    assert false, "cannot preserve symbol binding through newLit"
  of nnkNone:
    result = newCall(ident"newNimNode", ident"nnkNone")
  of nnkCommentStmt:
    result = newCall("newCommentStmtNode", newLit(n.strVal))
  of LitKinds:
    result = newCall(bindSym"newLit", n)
  else:
    # some nodes kinds have constructor procs
    case n.kind
    of nnkStmtList:
      result = newCall(ident"newStmtList")
    of nnkCall:
      result = newCall(ident"newCall")
    of nnkAsgn:
      result = newCall(ident"newAssignment")
    of nnkDotExpr:
      result = newCall(ident"newDotExpr")
    of nnkExprColonExpr:
      result = newCall(ident"newColonExpr")
    else:
      result = newCall(nnkDotExpr.newTree(ident($n.kind), ident"newTree"))
    for i in 0 ..< n.len:
      result.add newLit(n[i])
```

Reply:
doesn't work, `ident` isn't defined yet.

Reply:
Oh, I see. Could we not forward declare or move ident around?